Fix Select-Object to approve UX and fix return a property named *

iSazonov opened the pull request:
1. Now ExcludeProperty imply Property = '*' to approve UX #2351
2. Don't return property named '*'  #2420

Close #2351 
Close #2420

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

iSazonov committed:
Add test for Select-Object with Property = '*'

iSazonov committed:
Fix Select-Object  with Property = '*'

1. Now ExcludeProperty imply Property = '*'
2. Don't return property named '*'

iSazonov commented:
@PowerShellTeam please review the code.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs:
```diff
@@ -414,7 +419,7 @@ private void ProcessParameter(MshParameter p, PSObject inputObject, List<PSNoteP
-            if (expressionResults.Count == 0)
+            if (expressionResults.Count == 0 && !ex.HasWildCardCharacters)
             {
                 //Commented out for bug 1107600
```
Could you please remove the commented-out code snippet here? It's useless and there is no way to figure out what `bug 1107600` is now :)

daxian-dbw replied:
And it would be better if you can add some comments here, explaining your purpose (allow `Select-Object -Property noexist-name` to return a `PSObject` with property `noexist-name`, unless `noexist-name` itself contains wildcards).
I find the implementation of `Select-Object` is so poorly commented (near to zero comment) that it becomes a pain to read the code. So let's add some when we make changes in this file.

iSazonov replied:
I generally see little comments in code and tests and thought it was best practice. I will make more comments.
I think @PowerShellTeam  needs to mark this moment in documentation: "make more comments in your code".

daxian-dbw replied:
> I think @PowerShellTeam needs to mark this moment in documentation: "make more comments in your code".

I absolutely agree! I will update [coding-conventions](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#coding-conventions) to include this.

daxian-dbw replied:
#Closed.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+        # Issue #2420
+	It "Select-Object with implicit Property = '*' don't return property named '*'"{
+		$results = [pscustomobject]@{Thing="thing1"} | Select-Object -ExcludeProperty thing 
+		$results.psobject.Properties.Item("*").Name | Should Be $null
```
> Item("*").Name

`.Name` seems not necessary here.

iSazonov replied:
Done

daxian-dbw replied:
#Closed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs:
```diff
@@ -337,6 +337,11 @@ private void ProcessExpressionParameter()
             {
                 _exclusionFilter = new MshExpressionFilter(ExcludeProperty);
+                if ((Property == null) || (Property.Length == 0))
+                {
```
This will result a breaking change -- the scirpt below runs successfully before this change; but with this change a bunch errors will be thrown out.
```powershell
$s = Get-Process -Id $pid | Select-Object -ExcludeProperty ProcessName -ExpandProperty Modules
```
However, I don't feel this would break any existing scripts practically. It's documented that `-ExcludeProperty` has no effect without `-Property`, so I don't think any existing script would only specify `-ExcludeProperty`.

Maybe we can analyze the powershell scripts on GitHub to get some data. We can borrow @lzybkr's [AnalyzeCurlUsage.ps1](https://gist.github.com/lzybkr/3cd091334355f381d1d6ee7acfad5a48).

iSazonov replied:
1.The current tests do not catch it. I think we should add this in tests.

2.There is no difference between:
```powershell
 $s = Get-Process -Id $pid | Select-Object -ExpandProperty Modules
$s[0] | fl *
```
and
```powershell
$s = Get-Process -Id $pid | Select-Object -ExcludeProperty FileName -ExpandProperty Modules
$s[0] | fl *
```
i.e.  ExpandProperty skip ExcludeProperty. So the change could be corrected without  breaking.

3.The cmdlet is very intuitive to the user UX and very controversial:
ExcludeProperty is skipped w/o Property
ExpandProperty skip ExcludeProperty
ExpandProperty with Property throw exception
Compare:
```powershell
1 | Select-Object
"1" | Select-Object
1 | Select-Object -Property *
"1" | Select-Object -Property *
```
and many more.
And it is a tricky code, and any even slightly change this code can easily break down. 

(In general I think this cmdlet should be replaced with a newer, which meets modern requirements to cmdlets.)

Jaykul replied:
I think you mean that it's very **un**intuitive 😕 

If the intention is (to maintain the current expected behavior) that `-ExpandProperty` ignores `-ExcludeProperty` wouldn't it be better to have these switches in separate parameter sets? I'd certainly rather have it be _obvious that they can't be used together_ than to allow them together but ignore one (or worse, produce runtime errors if they're used together)...

iSazonov replied:
> I think you mean that it's very unintuitive 😕 

Under "intuitive" I mean that we use the cmdlet 10 years "as is" without many complaints (In other words, users get results they expect). 
All completely different if you look on the inner side of the cmdlet. I guess it was one of the first cmdlets of Powershell. Today such a design would be immediately rejected. The code is tricky and any chip change results in multiple test crush.
And we absolutely cannot add parameter sets (and moreover we cannot add new exceptions) because this is _breaking change_. 
As I said above, time to replace this cmdlet to a new one with stringent modern design, better UX  and better performance.

iSazonov replied:
@daxian-dbw I fix breaking change and add your test.

daxian-dbw replied:
I analyzed the usages of `Select-Object` in GitHub to find the usages like `Select-Object -ExcludeProperty ProcessName -ExpandProperty Modules`, which might be broken by this fix. The analysis scripts and results can be found [here](https://gist.github.com/daxian-dbw/07847ff5ef6da2c100e8b22bd94f0feb).

1074 .ps1/.psm1 files were found in GitHub with `Select-Object` in them, and totally 3376 instances of `Select-Object` were found. Out of them, 2 usages might be broken by the fix on `-ExcludeProperty`:
```
## FIND 2 usages like 'Select-Object -ExcludeProperty value -ExpandProperty value'
PS:96> $results | ? BreakingCase

UsesSelectObject     : True
FileName             : F:\temp\selectobject-dir-psm1\ECSNamespace.psm1
NeedManualInspection : False
BreakingCase         : True
Line                 : 48

UsesSelectObject     : True
FileName             : F:\temp\selectobject-dir-psm1\ECSvdc.psm1
NeedManualInspection : False
BreakingCase         : True
Line                 : 158
```
[ECSNamespace.psm1](https://raw.githubusercontent.com/bottkars/ECShell/6b2531a84e5107ccb1c6ab71a18e766f057efa14/Multitenancy/ECSNamespace.psm1) has `Select-Object -ExpandProperty $Expandproperty -ExcludeProperty $Excludeproperties` in it:
```
Invoke-RestMethod -Uri $Uri -Headers $ECSAuthHeaders -Method Get  -ContentType $ContentType | Select-Object -ExpandProperty $Expandproperty -ExcludeProperty $Excludeproperties | Select-Object $Properties #,@{N="NamespaceID";E={$_.id}}
```
[ECSvdc.psm1](https://raw.githubusercontent.com/bottkars/ECShell/6b2531a84e5107ccb1c6ab71a18e766f057efa14/Provisioning/ECSvdc.psm1) has `Select-Object -ExpandProperty vdc -ExcludeProperty $Excludeproperty` in it:
```
Invoke-RestMethod -Uri $Uri -Headers $ECSAuthHeaders -Method $Method -ContentType $ContentType  | Select-Object -ExpandProperty $Expandproperty | Select-Object -ExpandProperty vdc -ExcludeProperty $Excludeproperty
```

Although we actually find the usages that might be broken by this fix, but those 2 usages seems wrong in the first place -- the author indented to exclude some properties, which actually doesn't happen. 

Given the data, I feel it's OK to have this "breaking change" introduced by this fix.

daxian-dbw replied:
To avoid this breaking change, we will have to do something like the following:
```
When -ExcludeProperty is specified
    #1 if -Property is specified
        always respect -ExcludeProperty
    #2 else if -ExpandProperty is specified
        Ignore -ExcludeProperty
    #3 else
        set Property to "*" and respect -ExcludeProperty
```
But this will again raise confusion about how `-ExcludeProperty` actually works (when and when not it takes effect), and probably someone will see `#2` as a bug too. Therefore, I'm more inclined to accept always having `#3` when `-Property` is not specified, though it would introduce the breaking change discussed above. 
@lzybkr @adityapatwardhan @vors, what are your opinions on this?

iSazonov replied:
@daxian-dbw I made refactoring ProcessExpressionParameter() to the logic, add your test, add comments.

vors replied:
I'd be in favour of non-breaking change, so I agree with pseudo-code @daxian-dbw proposed.
Although it's still far from the clean design, it's a) non-breaking for existing legit usages b) better than what we currently have. So I'm considering it a reasonable step in the right direction.

iSazonov replied:
@vors Thanks!
@daxian-dbw I have implemented your suggestion in the last commit. Waiting for your conclusion.

iSazonov replied:
@daxian-dbw Thanks for summing-up. Await @lzybkr.
And my question remains whether to announce this cmdlet depredicated and design a new cmdlet with simple clear logic and easy to support code?

daxian-dbw replied:
@lzybkr this discussion thread was auto-collapsed in the main PR page, but you can read it at https://github.com/PowerShell/PowerShell/pull/2421/files/70c30384c74d9f6392b13cc9d04baf9eb5f8c364

iSazonov replied:
@daxian-dbw @lzybkr your conclusion?

daxian-dbw replied:
Sorry @iSazonov for the long delay.
@lzybkr and I had a quick discussion about this, and I think @lzybkr agreed that we should take the original fix with the *unlikely* breaking change.

iSazonov replied:
@daxian-dbw  Thanks! Good news!
Only I seem to be lost. What are my next steps? 
It seems I made all the changes you want in the code.

daxian-dbw replied:
The conclusion is to take your changes to `ProcessExpressionParameter()` in your second commit -- always add "*" to `-Property` when `-ExcludeProperty` is specified but `-Property` is not : https://github.com/PowerShell/PowerShell/pull/2421/commits/70c30384c74d9f6392b13cc9d04baf9eb5f8c364

I think you implemented the pseudo code I listed above to avoid the breaking change, but as we discussed here, that will raise new confusion about how `-ExlcudeProperty` works. So I believe it's reasonable to take your original changes to `ProcessExpressionParameter()` to make it simple for users to understand.

lzybkr replied:
Note that I hadn't actually concluded anything after my discussion w/ @daxian-dbw, but after some of my own experiments, I do agree. The existing behavior is surprising and not useful, anyone potentially hit by the breaking change wasn't getting what they wanted anyway.

iSazonov replied:
@daxian-dbw @lzybkr Thanks for the clarification and the final conclusion! I'll make the rollback.

iSazonov replied:
Done.

daxian-dbw replied:
Thank you @iSazonov 
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+        # Issue #2420
+	It "Select-Object with explicit Property = '*' don't return property named '*'"{
+		$results = [pscustomobject]@{Thing="thing1"} | Select-Object -Property * -ExcludeProperty thing 
+		$results.psobject.Properties.Item("*").Name | Should Be $null
```
`.Name` seems not necessary.

iSazonov replied:
Done

daxian-dbw replied:
#Closed.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+        # Issue #2351
+	It "Select-Object with implicit Property = '*' exclude not single prperty"{
+		$results = [pscustomobject]@{Thing="thing1";Param2="param2"} | Select-Object -ExcludeProperty Param2 
+                $results.Param2 | Should Be $null
```
Please fix the indentation here.

iSazonov replied:
Done

daxian-dbw replied:
#Closed.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+        # Issue #2351
+	It "Select-Object with explicit Property = '*' exclude not single prperty"{
+		$results = [pscustomobject]@{Thing="thing1";Param2="param2"} | Select-Object -Property * -ExcludeProperty Param2 
+                $results.Param2 | Should Be $null
```
Please fix the indentation here.

iSazonov replied:
Done

daxian-dbw replied:
#Closed.

RichardSiddaway commented:
Having separate parameter sets would be a great idea

Sent from my iPad

On 9 Oct 2016, at 20:30, Joel Bennett <notifications@github.com<mailto:notifications@github.com>> wrote:


@Jaykul commented on this pull request.

________________________________

In src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs<https://github.com/PowerShell/PowerShell/pull/2421>:

> @@ -337,6 +337,11 @@ private void ProcessExpressionParameter()
             if (ExcludeProperty != null)
             {
                 _exclusionFilter = new MshExpressionFilter(ExcludeProperty);
+                if ((Property == null) || (Property.Length == 0))
+                {


I think you mean that it's very unintuitive ?

If the intention is (to maintain the current expected behavior) that -ExpandProperty ignores -ExcludeProperty wouldn't it be better to have these switches in separate parameter sets? I'd certainly rather have it be obvious that they can't be used together than to allow them together but ignore one (or worse, produce runtime errors if they're used together)...

-
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/PowerShell/PowerShell/pull/2421>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHgYUaGr7eou2GSu_7Q0m2-iosO0xvc0ks5qyUDjgaJpZM4KNXvs>.

iSazonov committed:
Fixes after code review

Restore ExpandProperty skip ExcludeProperty
Add new test
Add new comments
Remove old comment

iSazonov commented:
@RichardSiddaway Sorry, this is the right decision but it is impossible to add parameter sets w/o breaking change. See my comment above.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs:
```diff
@@ -334,9 +334,11 @@ private void ProcessExpressionParameter()
 
-            if (ExcludeProperty != null)
+            // ExpandProperty skip processing ExcludeProperty
+            if (ExcludeProperty != null && string.IsNullOrEmpty(ExpandProperty))
```
We cannot ignore `-ExcludeProperty` just because `-ExpandProperty` is specified. See the script below:
```
PS:88> $s = Get-Process -Id $pid | Select-Object -Property Process* -ExpandProperty Modules
PS:89> $s[0].ProcessName
powershell
PS:90> $s[0].ProcessorAffinity
255
PS:91> $p = Get-Process -Id $pid | Select-Object -Property Process* -ExcludeProperty ProcessorAffinity
-ExpandProperty Modules
PS:92> $p[0].ProcessName
powershell
PS:93> $p[0].ProcessorAffinity
PS:94>
```
`-ExcludeProperty` took effect and excluded the property `ProcessorAffinity`, so only `ProcessName` got added to each Module object.

I feel it's OK to break the `Select-Object -ExcludeProperty value1 -ExpandProperty value2` usage based on my analysis of the usage of `Select-Object` in GitHub. Please see my comment above.

iSazonov replied:
Ok, waiting team approve above then revert the change.

iSazonov replied:
While waiting I made refactoring. See comment above.

iSazonov replied:
Done.

daxian-dbw replied:
#Closed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs:
```diff
@@ -419,18 +421,10 @@ private void ProcessParameter(MshParameter p, PSObject inputObject, List<PSNoteP
                 }
             }
 
+            // allow Select-Object -Property noexist-name to return a PSObject with property noexist-name,
```
It would be clearer if you add quotes around `Select-Object -Property noexist-name`.

iSazonov replied:
Done.

daxian-dbw replied:
#Closed

iSazonov committed:
Refactoring ProcessExpressionParameter()

1. Refactoring ProcessExpressionParameter() to make explicit logic for
processing Property, ExcludeProperty and ExpandProperty
2. Add new test
3. Correct comment

iSazonov committed:
Revert ProcessExpressionParameter() to second commit

1. Revert ProcessExpressionParameter() to second commit
2. Remove unneeded test

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,51 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+	}
+
+        # Issue #2351
+	It "Select-Object with implicit Property = '*' exclude not single prperty"{
```
> prperty

A typo, should be 'property'

iSazonov replied:
Fixed.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,51 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+	}
+
+        # Issue #2351
+	It "Select-Object with explicit Property = '*' exclude not single prperty"{
```
> prperty

Same typo :)

iSazonov replied:
It's fate :)
Fixed.

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1:
```diff
@@ -248,3 +248,51 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
+		$p = Get-Process -Id $pid | Select-Object -Property Process* -ExcludeProperty ProcessorAffinity -ExpandProperty Modules
+		$p[0].psobject.Properties.Item("ProcessorAffinity") | Should Be $null
+    }
+}
```
Can you please add a new line?

iSazonov replied:
Fixed.

daxian-dbw replied:
Thanks. #Closed

daxian-dbw commented:
Thanks @iSazonov for the updates! Once the typos in tests are fixed, the PR will be ready to merge.

iSazonov committed:
Fix typo in test

iSazonov committed:
Fix indentation

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs:
```diff
@@ -337,7 +339,13 @@ private void ProcessExpressionParameter()
+                if ((Property == null) || (Property.Length == 0))
+                 {
+                    Property = new Object[]{"*"};
+                    _propertyMshParameterList = processor.ProcessParameters(Property, invocationContext);                }
```
Done.

daxian-dbw reviewed (approved):
LGTM.

iSazonov commented:
@daxian-dbw Thanks very much!

daxian-dbw commented:
@iSazonov Thank you for your patience and the great contribution 👍
